File: Leetcode/tree/98_Validate Binary Search Tree/Solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        return self.dfs(root, -sys.maxint-1, sys.maxint)

    def dfs(self, root, lowerBound, higherBound):
        if not root:
            return True
        res = root.val>lowerBound and root.val<higherBound
        return res and self.dfs(root.left, lowerBound, root.val) and self.dfs(root.right, root.val, higherBound)

File: Leetcode/tree/98_Validate Binary Search Tree/Solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
private:
    bool dfs(TreeNode* root, long long lowerBound, long long higherBound){
        if(!root)
            return true;
        bool res = (root->val>lowerBound)&&(root->val<higherBound);
        return res && dfs(root->left, lowerBound, root->val) && dfs(root->right, root->val, higherBound);
    }
public:
    bool isValidBST(TreeNode* root) {
        return dfs(root, LLONG_MIN, LLONG_MAX);
    }
};
